File: internal/wire/wire_golden_test.go
// Copyright 2026 The go-hbase Authors
// SPDX-License-Identifier: Apache-2.0

package wire_test

import (
	"bytes"
	"os"
	"path/filepath"
	"runtime"
	"testing"

	"github.com/virogg/go-hbase/internal/wire"
	"github.com/virogg/go-hbase/internal/wiregolden"
)

func TestGoldenCorpus(t *testing.T) {
	dir := goldenDir(t)
	tsv, err := os.Open(filepath.Join(dir, "fixtures.tsv"))
	if err != nil {
		t.Fatalf("open fixtures.tsv: %v", err)
	}
	t.Cleanup(func() { _ = tsv.Close() })
	fixtures, err := wiregolden.Parse(tsv)
	if err != nil {
		t.Fatalf("parse fixtures.tsv: %v", err)
	}
	if len(fixtures) == 0 {
		t.Fatalf("no fixtures parsed")
	}

	for _, fx := range fixtures {
		t.Run(fx.Name, func(t *testing.T) {
			want, err := os.ReadFile(filepath.Join(dir, fx.Name+".bin"))
			if err != nil {
				t.Fatalf("read %s.bin: %v", fx.Name, err)
			}

			// (1) Encode → bytes
			var got bytes.Buffer
			opts := []wire.EncoderOption{}
			if fx.ChunkSize > 0 {
				opts = append(opts, wire.WithChunkSize(fx.ChunkSize))
			}
			msg := fx.Message
			if err := wire.NewEncoder(&got, opts...).Encode(&msg); err != nil {
				t.Fatalf("Encode: %v", err)
			}
			if !bytes.Equal(got.Bytes(), want) {
				t.Fatalf("encode bytes mismatch:\n  got  (%d): %x\n  want (%d): %x",
					got.Len(), got.Bytes(), len(want), want)
			}

			// (2) bytes → Decode
			decoded, err := wire.NewDecoder(bytes.NewReader(want)).Decode()
			if err != nil {
				t.Fatalf("Decode: %v", err)
			}
			if decoded.Type != fx.Message.Type ||
				decoded.ReqID != fx.Message.ReqID ||
				decoded.RegionID != fx.Message.RegionID ||
				decoded.HookID != fx.Message.HookID ||
				!bytes.Equal(decoded.Payload, normalizePayload(fx.Message.Payload)) {
				t.Fatalf("decoded mismatch:\n  got  %+v\n  want %+v", decoded, fx.Message)
			}
		})
	}
}

func goldenDir(t *testing.T) string {
	t.Helper()
	_, file, _, ok := runtime.Caller(0)
	if !ok {
		t.Fatalf("runtime.Caller failed")
	}
	return filepath.Join(filepath.Dir(file), "..", "..", "test", "golden", "wire", "v1")
}

func normalizePayload(p []byte) []byte {
	if p == nil {
		return []byte{}
	}
	return p
}

File: test/java/com/virogg/hbasecop/bridge/wire/WireGoldenTest.java
// Copyright 2026 The go-hbase Authors
// SPDX-License-Identifier: Apache-2.0

package com.virogg.hbasecop.bridge.wire;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.MethodSource;

/**
 * Java half of the T13 cross-language byte-level verification. Pairs with
 * internal/wire/wire_golden_test.go on the Go side: both load the same fixtures.tsv and the same
 * {@code <name>.bin} files, then assert that encode→bytes and bytes→decode produce identical
 * results.
 *
 * <p>Test working directory at runtime is the Maven project basedir, so the fixtures path is
 * relative to repo root.
 */
class WireGoldenTest {

  private static final Path CORPUS_DIR = Paths.get("test/golden/wire/v1");

  static List<Fixture> fixtures() throws IOException {
    List<String> lines = Files.readAllLines(CORPUS_DIR.resolve("fixtures.tsv"));
    List<Fixture> out = new ArrayList<>();
    for (String raw : lines) {
      String line = raw.trim();
      if (line.isEmpty() || line.startsWith("#")) {
        continue;
      }
      String[] f = line.split("\\|", -1);
      if (f.length != 8) {
        throw new IllegalStateException("expected 8 fields, got " + f.length + ": " + line);
      }
      FrameType type = FrameType.valueOf(f[1]);
      long reqId = Long.parseUnsignedLong(f[2]);
      int regionId = Integer.parseUnsignedInt(f[3]);
      byte hookId = (byte) Integer.parseUnsignedInt(f[4]);
      byte[] payload = parsePayload(f[5], f[6]);
      int chunkSize = Integer.parseInt(f[7]);
      out.add(new Fixture(f[0], new Message(type, reqId, regionId, hookId, payload), chunkSize));
    }
    return out;
  }

  @ParameterizedTest(name = "[{index}] {0}")
  @MethodSource("fixtures")
  @DisplayName("encode→.bin matches; .bin→decode matches")
  void goldenRoundTrip(Fixture fx) throws IOException {
    byte[] want = Files.readAllBytes(CORPUS_DIR.resolve(fx.name + ".bin"));

    // (1) Encode → bytes
    Encoder enc = fx.chunkSize > 0 ? new Encoder(fx.chunkSize) : new Encoder();
    ByteBuffer encoded = enc.encode(fx.message);
    byte[] got = new byte[encoded.remaining()];
    encoded.get(got);
    assertArrayEquals(want, got, fx.name + ": encode bytes mismatch");

    // (2) bytes → Decode
    Message decoded = new Decoder().decode(ByteBuffer.wrap(want));
    assertEquals(fx.message.type(), decoded.type(), fx.name + ": type");
    assertEquals(fx.message.reqId(), decoded.reqId(), fx.name + ": reqId");
    assertEquals(fx.message.regionId(), decoded.regionId(), fx.name + ": regionId");
    assertEquals(fx.message.hookId(), decoded.hookId(), fx.name + ": hookId");
    assertArrayEquals(fx.message.payload(), decoded.payload(), fx.name + ": payload");
  }

  private static byte[] parsePayload(String kind, String value) {
    switch (kind) {
      case "HEX":
        return parseHex(value);
      case "ASCENDING":
        int n = Integer.parseInt(value);
        byte[] out = new byte[n];
        for (int i = 0; i < n; i++) {
          out[i] = (byte) (i % 251);
        }
        return out;
      default:
        throw new IllegalArgumentException("unknown payload_kind: " + kind);
    }
  }

  // Hand-rolled hex parser; java.util.HexFormat is Java 17+, we target 11.
  private static byte[] parseHex(String hex) {
    if (hex.isEmpty()) {
      return new byte[0];
    }
    if ((hex.length() & 1) != 0) {
      throw new IllegalArgumentException("odd-length hex: " + hex);
    }
    byte[] out = new byte[hex.length() / 2];
    for (int i = 0; i < out.length; i++) {
      int hi = Character.digit(hex.charAt(i * 2), 16);
      int lo = Character.digit(hex.charAt(i * 2 + 1), 16);
      if (hi < 0 || lo < 0) {
        throw new IllegalArgumentException("non-hex char near index " + (i * 2));
      }
      out[i] = (byte) ((hi << 4) | lo);
    }
    return out;
  }

  // Sanity assertion that the corpus directory is reachable from the
  // Maven working directory. Catches accidental cwd drift before any
  // parameterized test fails with a less obvious file-not-found.
  @org.junit.jupiter.api.Test
  void corpusDirectoryExists() {
    assertFalse(
        !Files.isRegularFile(CORPUS_DIR.resolve("fixtures.tsv")),
        "missing " + CORPUS_DIR.resolve("fixtures.tsv").toAbsolutePath());
  }

  static final class Fixture {
    final String name;
    final Message message;
    final int chunkSize;

    Fixture(String name, Message message, int chunkSize) {
      this.name = name;
      this.message = message;
      this.chunkSize = chunkSize;
    }

    @Override
    public String toString() {
      return name;
    }
  }
}
